File: TownOfSalemCalculator/src/townofsalemcalculator/Conditions/Abstract/AdvancedConditions/AllUniqueness.java
package townofsalemcalculator.Conditions.Abstract.AdvancedConditions;

import townofsalemcalculator.Conditions.Abstract.BasicConditions.RoleUniqueness;
import townofsalemcalculator.Conditions.Abstract.InheritConditions.CompositeCondition;
import static townofsalemcalculator.Role.*;

/**
 * All roles that are unique are included in this condition
 * @author Multifacio
 * @version 1.0
 * @since 2018-10-30
 */
public final class AllUniqueness extends CompositeCondition {
    public AllUniqueness() {
        addCondition(new RoleUniqueness(Jailor));
        addCondition(new RoleUniqueness(Mayor));
        addCondition(new RoleUniqueness(Retributionist));
        addCondition(new RoleUniqueness(Veteran));
        addCondition(new RoleUniqueness(Godfather));
        addCondition(new RoleUniqueness(Mafioso));
        addCondition(new RoleUniqueness(Werewolf));
    }
}

File: src/Conditions/Start/AllUniqueness.py
from src.Concepts.Role import Role
from src.Conditions.Abstract.ANDCondition import ANDCondition
from src.Concepts.Rolegroup import Rolegroup as RG
from src.Conditions.Abstract.AtMostCondition import AtMostCondition

class AllUniqueness(ANDCondition):
    """ The All Uniqueness Condition enforces that all unique roles should occur at most once. """
    UNIQUE_ROLES = frozenset({Role.JAILOR, Role.VETERAN, Role.MAYOR, Role.RETRIBUTIONIST, Role.GODFATHER, Role.MAFIOSO,
                              Role.WEREWOLF})
    def __init__(self):
        super().__init__([AtMostCondition(frozenset({role}), 1) for role in self.UNIQUE_ROLES])